picoquic/sender.c

```diff
@@ -1469,9 +1474,23 @@ int picoquic_prepare_packet_ready(picoquic_cnx_t* cnx, picoquic_packet* packet,
             }
 
             if (cnx->cwin > cnx->bytes_in_transit) {
+                /* If keep alive is enabled and it is time to send the next frame, send it! */
+                if (cnx->keep_alive != NULL && cnx->keep_alive->next_frame_time <= current_time) {
+                    ret = picoquic_prepare_misc_frame(cnx->keep_alive->frame, &bytes[length],
+                                                      cnx->send_mtu - checksum_overhead - length, &data_bytes);
+                    if (ret == 0) {
+                        length += data_bytes;
+                        cnx->keep_alive->next_frame_time = current_time + cnx->keep_alive->interval;
+                    } else {
+                        if (ret == PICOQUIC_ERROR_FRAME_BUFFER_TOO_SMALL) {
+                            ret = 0;
+                        }
+                    }
```

Comment:
This could be simplified if you go for the "simple ping" approach. You would just need something like:
~~~
if (cnx->keep_alive != NULL && cnx->keep_alive->next_frame_time <= current_time && length  + checksum_overhead + 1< cnx->send_mtu) {
    bytes[length++] = picoquic_frame_type_ping;
    bytes[length++] = 0;
    cnx->keep_alive->next_frame_time = current_time + cnx->keep_alive->interval;
}
~~~
No need to involve the "misc frame type" path -- it is really meant for frames directly formatted by the application.

Reply:
Yeah, got catch!